### PR TITLE
Adds targeted functionality to hailers

### PR DESCRIPTION
**Pull request by 9600bauds**
Solves #9244
What this does is let you click on people to hail them directly. This creates a much more noticeable message on their end:
dreamseeker_2016-05-27_18-07-34
As you can see, it also writes in chat who is hailing who, so all bystanders can know what happened.
It will also draw a warning on the suspect, and a little effect on the officer to make him easier to identify. 
mspaint_2016-06-02_20-37-58
The images last for 2 seconds.
There's a bit of "aim assist" here to help you hail moving targets: If you didn't manage to click anyone directly, then it will try to hail every mob in the tile you clicked. If there are no mobs, it will hail everyone in a 3x3 area around your cursor.

### CODE REVIEW

File: html/changelogs/9600bauds_hhhh.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+# Optional: Remove this file after generating master changelog.  Useful for PR changelogs that won't get used again.
+delete-after: True
+
+# Any changes you've made.  See valid prefix list above.
+# There needs to be a space after the - and before the prefix. Don't use tabs anywhere in this file.
+# Also, this gets changed to [] after reading.  Just remove the brackets when you add new shit.
+# If you're using characters such as # ' : - or the like in your changelog, surround the entire change with quotes as shown in the second example. These quotes don't show up on the changelog once it's merged and prevent errors.
+# SCREW ANY OF THIS UP AND IT WON'T WORK.
+changes: 
+- bugfix: "Fixed speech bubbles not changing their icon when your speech ends in '?' or '!'."
```
Comment (9600bauds) [inline html/changelogs/9600bauds_hhhh.yml]: fug I branched this this over the branch I used for #10220

File: code/game/objects/items/devices/whistle.dm [hunk 2]
```diff
@@ -13,32 +13,90 @@
...
-								"<span class='warning'>Your [name] gurgles, \"FUCK YOUR CUNT YOU SHIT EATING COCKSUCKER MAN EAT A DONG FUCKING ASS RAMMING SHITFUCK. EAT PENISES IN YOUR FUCKFACE AND SHIT OUT ABORTIONS OF FUCK AND DO A SHIT IN YOUR ASS YOU COCK FUCK SHIT MONKEY FUCK ASS WANKER FROM THE DEPTHS OF SHIT.\"</span>", \
-								"<span class='warning'>You hear the computerized voice of a security hailer: \"FUCK YOUR CUNT YOU SHIT EATING COCKSUCKER MAN EAT A DONG FUCKING ASS RAMMING SHITFUCK. EAT PENISES IN YOUR FUCKFACE AND SHIT OUT ABORTIONS OF FUCK AND DO A SHIT IN YOUR ASS YOU COCK FUCK SHIT MONKEY FUCK ASS WANKER FROM THE DEPTHS OF SHIT.\"</span>")
-			insults--
-		else
-			to_chat(user, "<span class='warning'>*BZZZZcuntZZZZT*</span>")
+		playsound(get_turf(src), 'sound/voice/binsult.ogg', 100, 1, vary = 0)//hueheuheuheuheuheuhe
+		user.visible_message("<span class='warning'>[user]'s [name] gurgles, \"[say_your_thing()]\"</span>", \
```
Comment (ComicIronic) [inline code/game/objects/items/devices/whistle.dm]: Don't call it multiple times, call it once and assign the value.

File: code/game/objects/items/devices/whistle.dm [hunk 3]
```diff
@@ -13,32 +13,90 @@
...
+	for(var/mob/living/M in suspects)
+		M.show_message("<span class='userdanger'>[bicon(src)][say_your_thing()]</span>", MESSAGE_HEAR)
+	var/who = suspects.len <= 3 ? english_list(suspects) : "everyone"
+	user.visible_message("<span class='danger'>[user] hails for [who] to halt!</span>", \
+						"<span class='warning'>You hail for [who] to halt!</span>", \
+						"<span class='warning'>You hear the computerized voice of a security hailer: \"[say_your_thing()]\"</span>")
+
+	// ~ sound ~ //
+	if(emagged)
+		playsound(get_turf(src), 'sound/voice/binsult.ogg', 100, 1, vary = 0)
```
Comment (ComicIronic) [inline code/game/objects/items/devices/whistle.dm]: Make this sound playing a proc as well, it shouldn't be repeated.

**Comment by 9600bauds** [comment]
This is what I have right now, which is shit
mspaint_2016-06-02_16-59-21
Please sprite me

**Comment by Blithering** [comment]
http://s000.tinyupload.com/index.php?file_id=79468562249415251029 These any good?

**Comment by 9600bauds** [comment]
mspaint_2016-06-02_17-31-58
Unfortunately does not play nice with sechud overlays but does look a little better without them
mspaint_2016-06-02_17-33-22

**Comment by Blithering** [comment]
http://s000.tinyupload.com/index.php?file_id=67165975501624342876 Slightly more sechud friendly version.

**Comment by 9600bauds** [comment]
mspaint_2016-06-02_20-37-58
Better

File: code/modules/clothing/head/helmet.dm [hunk 4]
```diff
@@ -27,6 +27,8 @@
 						"<span class='warning'>Your [name] rasps, \"WOOP WOOP!\"</span>", \
 						"<span class='warning'>You hear a siren: \"WOOP WOOP!\"</span>")
 
+	var/list/bystanders = get_hearers_in_view(world.view, src)
+	flick_overlay(image('icons/mob/talk.dmi', user, "hail", MOB_LAYER+1), clients_in_moblist(bystanders), 15)
```
Comment (9600bauds) [inline code/modules/clothing/head/helmet.dm:31]: stealth woop buff

**Comment by SangriaMonkey** [comment]
People are still going to tear ass in the other direction the second they hear it

**Comment by 9600bauds** [comment]
Now they can't claim they didn't hear it, at least

File: code/modules/mob/living/say.dm [hunk 5]
```diff
@@ -246,7 +246,7 @@ var/list/department_radio_keys = list(
 	for (var/atom/movable/listener in listeners)
 		listener.Hear(speech, rendered)
 
-	send_speech_bubble(speech, bubble_type, listeners)
+	send_speech_bubble(speech.message, bubble_type, listeners)
```
Comment (9600bauds) [inline code/modules/mob/living/say.dm:249]: still leftover from #10220 but has no effect

**Comment by Blithering** [comment]
Alternatively, offset the exclamation mark to above the head and play the mgs alert noise when targetted.

**Comment by MercuryDoll** [comment]
> Alternatively, offset the exclamation mark to above the head and play the mgs alert noise when targetted.

This. Also make the exclamation mark slightly more brighter/saturated for better noticability.

**Comment by ghost** [comment]
This is a buff to the best item in the game.

**Comment by 9600bauds** [comment]
The exclamation mark blinks so it's decently visible as is

**Comment by Killette2** [comment]
Thumb this down if you are Irish.

**Comment by 9600bauds** [comment]
Any words against?

**Comment by ihadtoregisterforthis** [comment]
If this were to be used on a MED RP round, how much would inmersion suffer?

**Comment by 9600bauds** [comment]
The hailer has a holo-projector, done.

**Comment by MercuryDoll** [comment]
Medium RP would still be fine because people should have an ability called "suspension of disbelief" that makes them able to incorporate that immersively into their experience. Only the Autismus Maximi on Heavy RP server would be crying bloody tears about it.
It's a greatly needed buff to the item, i'd say merge it.

File: code/__HELPERS/game.dm [hunk 6]
```diff
@@ -350,9 +350,9 @@ var/list/DummyCache = list()
 /proc/flick_overlay(image/I, list/show_to, duration)
 	for(var/client/C in show_to)
 		C.images += I
-	sleep(duration)
-	for(var/client/C in show_to)
-		C.images -= I
+	spawn(duration)
```
Comment (PJB3005) [inline code/__HELPERS/game.dm]: Give the proc `waitfor = FALSE` instead.

Comment (9600bauds) [inline code/__HELPERS/game.dm]: What's that?

Comment (clusterfack) [inline code/__HELPERS/game.dm]: It turns sleeps into spawns

File: code/game/objects/items/devices/whistle.dm [hunk 7]
```diff
@@ -12,33 +12,84 @@
...
+	if(emagged && insults)
+		playsound(get_turf(src), 'sound/voice/binsult.ogg', 100, 1, vary = 0)
+		insults--
+	else
+		playsound(get_turf(src), 'sound/voice/halt.ogg', 100, 1, vary = 0)
+	if(user)
+		var/list/bystanders = get_hearers_in_view(world.view, src)
+		flick_overlay(image('icons/mob/talk.dmi', user, "hail", MOB_LAYER+1), clients_in_moblist(bystanders), 2 SECONDS)
 	spamcheck = 1
 	spawn(20)
```
Comment (clusterfack) [inline code/game/objects/items/devices/whistle.dm]: Change this to nextuse = world.time + cooldown, and then check elsewhere if world.time > nextuse. Spawn is pretty crap for this sort of thing.

**Comment by clusterfack** [comment]
Other than mine and pjb's comment this is ready for merge

**Comment by 9600bauds** [comment]
Tested the last commit and werks

**Comment by nuklearcellphoneg** [comment]
Just to be sure, you did add the MGS noise? :^)

**Pull request merged**